### PR TITLE
Harvesting, downloading and hazardset completion processes

### PR DESCRIPTION
**Pull request by arnaud-morvan**
This PR adds:
- harvesting from geonode
- downloading raster data
- completing hazardsets

Does not yet react to medata changes, but comments are welcome.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -133,6 +133,19 @@ Processing tasks
 Thinkhazard_processing provides several consecutive tasks to populate the
 thinkhazard datamart database. These are:
 
+``.build/venv/bin/harvest [--force] [--dry-run]``
+
+Havest metadata from geonode, create HazardSet and Layer records.
```
Comment (fvanderbiest) [inline README.rst]: Harvest

File: README.rst [hunk 2]
```diff
@@ -133,6 +133,19 @@ Processing tasks
...
+
+``.build/venv/bin/download [--title] [--force] [--dry-run]``
+
+Download raster files in data folder.
+
+``.build/venv/bin/complete [--force] [--dry-run]``
+
+Check hazardsets that are complete, calculate some fields and pin them as
```
Comment (fvanderbiest) [inline README.rst]: Identify hazardsets whose layers have been fully downloaded, ...

File: thinkhazard_processing/completing.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                transaction.commit()
+        except Exception as e:
+            transaction.abort()
+            logger.error(e.message)
+
+
+def complete_hazardset(hazardset_id, dry_run=False):
+    hazardset = DBSession.query(HazardSet).get(hazardset_id)
+    if hazardset is None:
+        raise Exception('HazardSet {} does not exist.'
```
Comment (fvanderbiest) [inline thinkhazard_processing/completing.py]: Hazardset

File: thinkhazard_processing/completing.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                        .format(hazardset_id))
+
+    hazardtype = hazardset.hazardtype
+    type_settings = settings['hazard_types'][hazardtype.mnemonic]
+    preprocessed = type_settings['preprocessed']
+
+    layers = []
+    if preprocessed:
+        if len(hazardset.layers) == 0:
+            logger.info('  No layer founded')
```
Comment (fvanderbiest) [inline thinkhazard_processing/completing.py]: found

File: thinkhazard_processing/completing.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+def complete_hazardset(hazardset_id, dry_run=False):
+    hazardset = DBSession.query(HazardSet).get(hazardset_id)
+    if hazardset is None:
+        raise Exception('HazardSet {} does not exist.'
+                        .format(hazardset_id))
+
+    hazardtype = hazardset.hazardtype
+    type_settings = settings['hazard_types'][hazardtype.mnemonic]
+    preprocessed = type_settings['preprocessed']
```
Comment (fvanderbiest) [inline thinkhazard_processing/completing.py]: will require an update

File: thinkhazard_processing/completing.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            transaction.abort()
+            raise
+
+    ids = DBSession.query(HazardSet.id)
+    if not force:
+        ids = ids.filter(HazardSet.complete.is_(False))
+    if hazardset_id is not None:
+        ids = ids.filter(HazardSet.id == hazardset_id)
+    for id in ids:
+        logger.info(id[0])
```
Comment (fvanderbiest) [inline thinkhazard_processing/completing.py]: logger.info('Working on hazardset {}'.format(id[0]))

File: thinkhazard_processing/completing.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        Layer.local,
+        func.min(Layer.data_lastupdated_date),
+        func.min(Layer.metadata_lastupdated_date),
+        func.min(Layer.calculation_method_quality),
+        func.min(Layer.scientific_quality)) \
+        .filter(Layer.hazardset_id == hazardset.id) \
+        .group_by(Layer.local)
+
+    if stats.count() > 1:
+        logger.warning('  Mixed local and global layers')
```
Comment (fvanderbiest) [inline thinkhazard_processing/completing.py:104]: This case should prevent the hazardset from being marked as complete

File: thinkhazard_processing/harvesting.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    if not hazard_type:
+        logger.warning('{} - hazard_type is empty'.format(title))
+        return False
+    hazardtype = hazardtype_from_geonode(hazard_type)
+    if hazardtype is None:
+        logger.warning('{} - hazard_type not supported'.format(title))
+        return False
+
+    type_settings = settings['hazard_types'][hazardtype.mnemonic]
+    preprocessed = type_settings['preprocessed']
```
Comment (fvanderbiest) [inline thinkhazard_processing/harvesting.py]: update required

File: thinkhazard_processing/downloading.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                          layer.download_url,
+                          '',
+                          ''))
+        logger.info('Retrieving {}'.format(url))
+        response, content = h.request(url)
+
+        with open(layer.path(), 'wb') as f:
+            f.write(content)
+
+    layer.downloaded = True
```
Comment (fvanderbiest) [inline thinkhazard_processing/downloading.py]: layer should be marked as downloaded on file write success

File: thinkhazard_processing/harvesting.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+            if dry_run:
+                transaction.abort()
+            else:
+                transaction.commit()
+        except:
+            transaction.abort()
+            raise
+
+    params = {}
+    if hazard_type is not None:
```
Comment (fvanderbiest) [inline thinkhazard_processing/harvesting.py:57]: Improvement: harvest all hazardtypes at once

File: thinkhazard_processing/harvesting.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+        params['hazard_type__in'] = hazard_type
+
+    hazard_type_url = urlunsplit((geonode['scheme'],
+                                  geonode['netloc'],
+                                  'api/layers/',
+                                  urlencode(params),
+                                  ''))
```
Comment (fvanderbiest) [inline thinkhazard_processing/harvesting.py:64]: Improvement: harvest recently updated metadata only, eg with &metadata_update_date__gte=2015-05-30T13:00:00

File: thinkhazard_processing/harvesting.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+    csw_wkt_geometry = object['csw_wkt_geometry']
+    bounds = wkt_loads(csw_wkt_geometry).bounds
+    # print '  bounds :', bounds
+
+    # TODO: minimum global bbox should be defined in settings
+    minimum_global_bounds = (-175, -45, 175, 45)
+    from shapely.geometry import box
+    local = not box(bounds).contains(box(minimum_global_bounds))
+    '''
+    local = 'GLOBAL' not in hazardset_id
```
Comment (fvanderbiest) [inline thinkhazard_processing/harvesting.py:121]: Indeed, TODO in another PR

File: thinkhazard_processing/harvesting.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+    if not scientific_quality:
+        logger.warning('{} - scientific_quality is empty'.format(title))
+        return False
+    scientific_quality = int(float(scientific_quality))
+
+    download_url = object['download_url']
+    if not download_url:
+        logger.warning('{} - download_url is empty'.format(title))
+        return False
+
```
Comment (fvanderbiest) [inline thinkhazard_processing/harvesting.py:196]: TODO: take into account `distribution_url`

File: thinkhazard_processing/harvesting.py [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+    hazardset = DBSession.query(HazardSet).get(hazardset_id)
+    if hazardset is None:
+
+        logger.info('{} - Create new HazardSet {}'
```
Comment (fvanderbiest) [inline thinkhazard_processing/harvesting.py]: Create new hazardset

File: thinkhazard_processing/scripts/complete.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+def main(argv=sys.argv):
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '-f', '--force', dest='force',
+        action='store_const', const=True, default=False,
+        help='Force execution even if dataset has already been processed')
+    parser.add_argument(
+        '--dry-run', dest='dry_run',
+        action='store_const', const=True, default=False,
+        help='Perform a trial run that do not commit changes')
```
Comment (fvanderbiest) [inline thinkhazard_processing/scripts/complete.py]: does

File: thinkhazard_processing/scripts/download.py [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+        '--title', dest='title', action='store',
+        help='The layer title')
+    parser.add_argument(
+        '--force', dest='force',
+        action='store_const', const=True, default=False,
+        help='Force download even if layer has already been download')
+    parser.add_argument(
+        '--dry-run', dest='dry_run',
+        action='store_const', const=True, default=False,
+        help='Perform a trial run that do not commit changes')
```
Comment (fvanderbiest) [inline thinkhazard_processing/scripts/download.py]: does

File: thinkhazard_processing/scripts/download.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+def main(argv=sys.argv):
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--title', dest='title', action='store',
+        help='The layer title')
+    parser.add_argument(
+        '--force', dest='force',
+        action='store_const', const=True, default=False,
+        help='Force download even if layer has already been download')
```
Comment (fvanderbiest) [inline thinkhazard_processing/scripts/download.py]: downloaded

File: thinkhazard_processing/scripts/harvest.py [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+from .. import settings
+from ..harvesting import harvest
+
+
+def main(argv=sys.argv):
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '-f', '--force', dest='force',
+        action='store_const', const=True, default=False,
+        help='Force execution even if dataset has already been processed')
```
Comment (fvanderbiest) [inline thinkhazard_processing/scripts/harvest.py]: harvested

File: thinkhazard_processing/scripts/harvest.py [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    parser.add_argument(
+        '--dry-run', dest='dry_run',
+        action='store_const', const=True, default=False,
+        help='Perform a trial run that do not commit changes')
```
Comment (fvanderbiest) [inline thinkhazard_processing/scripts/harvest.py]: does

File: Makefile [hunk 20]
```diff
@@ -13,6 +13,7 @@ help:
 	@echo "- initdb                  Initialize database"
 	@echo "- check                   Check the code with flake8"
 	@echo "- test                    Run the unit tests"
+	@echo "- harvest                 Harvest geonode"
```
Comment (fvanderbiest) [inline Makefile]: missing download & complete

File: thinkhazard_processing/completing.py (hunk 7 above)
Comment (fvanderbiest) [inline thinkhazard_processing/completing.py:104]: done

**Comment by pgiraud** [comment]
I trust your review and fixes. Please merge when travis is OK.

**Pull request merged**